### PR TITLE
gh-67877: Fix memory leaks in terminated RE matching

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
If SRE(match) function terminates abruptly, either because of a signal or because memory allocation fails, allocated SRE_REPEAT blocks might be never released.

* Issue: gh-67877

### CODE REVIEW

**Comment by serhiy-storchaka** [comment]
See also  #126843.

File: Modules/_sre/sre.c [hunk 1]
```diff
@@ -502,6 +581,11 @@ state_init(SRE_STATE* state, PatternObject* pattern, PyObject* string,
     state->must_advance = 0;
     state->debug = ((pattern->flags & SRE_FLAG_DEBUG) != 0);
 
+    /* SRE_REPEAT pool */
+    state->repeat = NULL;
+    state->repeat_pool_used = NULL;
+    state->repeat_pool_unused = NULL;
+
```
Comment (wjssz) [inline Modules/_sre/sre.c]: I'm sorry, above code already zeroed the state object: `memset(state, 0, sizeof(SRE_STATE));`
So here can be replaced by `assert()`:
```c
    /* SRE_REPEAT pool */
    assert(state->repeat == NULL);
    assert(state->repeat_pool_used == NULL);
    assert(state->repeat_pool_unused == NULL);
```
If you like, these two assignments can use `assert()` as well:
```c
    assert(state->match_all == 0);
    assert(state->must_advance == 0);
```

File: Modules/_sre/sre.c [hunk 2]
```diff
@@ -1506,6 +1631,11 @@ _sre_compile_impl(PyObject *module, PyObject *pattern, int flags,
     self->pattern = NULL;
     self->groupindex = NULL;
     self->indexgroup = NULL;
+#ifdef Py_DEBUG
+    self->fail_after_count = -1;
+    self->fail_after_exc = NULL;
+    self->fail_after_exc = Py_NewRef(PyExc_RuntimeError);
```
Comment (wjssz) [inline Modules/_sre/sre.c]: This line can be removed:
`self->fail_after_exc = Py_NewRef(PyExc_RuntimeError);`

File: Modules/_sre/sre.c [hunk 3]
```diff
@@ -2826,7 +2956,8 @@ pattern_scanner(_sremodulestate *module_state,
         return NULL;
     }
 
-    scanner->pattern = Py_NewRef(self);
+    Py_INCREF(self);
+    scanner->pattern = self;
```
Comment (wjssz) [inline Modules/_sre/sre.c:2951]: Why change here?
`Py_NewRef()` was added in Python 3.10.

File: Modules/_sre/sre.c [hunk 4]
```diff
@@ -533,6 +622,13 @@ state_fini(SRE_STATE* state)
     /* See above PyMem_Free() for why we explicitly cast here. */
     PyMem_Free((void*) state->mark);
     state->mark = NULL;
+    /* SRE_REPEAT pool */
+    repeat_pool_clear(state);
+#ifdef Py_DEBUG
+    if (pattern) {
+        pattern->fail_after_count = -1;
+    }
+#endif
```
Comment (wjssz) [inline Modules/_sre/sre.c]: How about remove `pattern->fail_after_count = -1;` here?
This counter is only for debuging purpose, maybe removing the resetting is irrelevant.

Then there won't be these changes everywhere in this patch:
```diff
-       state_fini(&state);
+       state_fini(&state, self);
```

File: Modules/_sre/sre.c [hunk 5]
```diff
@@ -1381,6 +1483,29 @@ _sre_SRE_Pattern___deepcopy__(PatternObject *self, PyObject *memo)
...
+
+static PyObject *
+_sre_SRE_Pattern__fail_after_impl(PatternObject *self, int count,
+                                  PyObject *exception)
+/*[clinic end generated code: output=9a6bf12135ac50c2 input=ef80a45c66c5499d]*/
+{
+    self->fail_after_count = count;
+    Py_INCREF(exception);
+    Py_XSETREF(self->fail_after_exc, exception);
+    Py_RETURN_NONE;
```
Comment (wjssz) [inline Modules/_sre/sre.c:1495]: Is it need to check `exception` variable with `PyExceptionClass_Check(exception)`?

If you remove the counter resetting in `state_fini()`, use `p._fail_after(-1, None)` to disable the functionality manually:
```c
    if (count < 0) {
        Py_CLEAR(self->fail_after_exc);
    } else {
        if (!PyExceptionClass_Check(exception)) {
            PyErr_SetString(PyExc_TypeError,
                            "expected exception type when count >= 0");
            return NULL;
        }
        Py_INCREF(exception);
        Py_XSETREF(self->fail_after_exc, exception);
    }
    self->fail_after_count = count;
    Py_RETURN_NONE;
```

File: Modules/_sre/sre.c (hunk 4 above)
Comment (serhiy-storchaka) [reply Modules/_sre/sre.c]: I feel safer if it is reset. But, well, I'll do this for the sake of minimizing the diff.

File: Modules/_sre/sre.c (hunk 3 above)
Comment (serhiy-storchaka) [reply Modules/_sre/sre.c:2951]: Because the type of ` scanner->pattern` was changed to `PatternObject*`.

File: Modules/_sre/sre.c (hunk 5 above)
Comment (serhiy-storchaka) [reply Modules/_sre/sre.c:1495]: I think this is an overcomplication. I have not added the check to able to call `_fail_after(-1, None)`. If you specify non-exception with non-negative count, you will just get a TypeError or SystemError when trying to raise it.

File: Modules/_sre/sre.c (hunk 3 above)
Comment (wjssz) [reply Modules/_sre/sre.c:2951]: I see.

File: Modules/_sre/sre.c (hunk 5 above)
Comment (wjssz) [reply Modules/_sre/sre.c:1495]: ok

File: Modules/_sre/sre_lib.h [hunk 6]
```diff
@@ -560,13 +560,29 @@ typedef struct {
...
 
+#ifdef Py_DEBUG
+# define MAYBE_CHECK_SIGNALS                                       \
+    do {                                                           \
+        _MAYBE_CHECK_SIGNALS;                                      \
+        if (state->fail_after_count >= 0) {                        \
+            if (state->fail_after_count == 0) {                    \
+                PyErr_SetNone(state->fail_after_exc);              \
+                RETURN_ERROR(SRE_ERROR_INTERRUPTED);               \
+            }                                                      \
```
Comment (wjssz) [inline Modules/_sre/sre_lib.h:578]: Maybe set `state->fail_after_count = -1` here is a good style?

File: Modules/_sre/sre_lib.h [hunk 7]
```diff
@@ -560,13 +560,28 @@ typedef struct {
...
             RETURN_ERROR(SRE_ERROR_INTERRUPTED);                   \
         }                                                          \
     } while (0)
 
+#ifdef Py_DEBUG
+# define MAYBE_CHECK_SIGNALS                                       \
+    do {                                                           \
+        _MAYBE_CHECK_SIGNALS;                                      \
+        if (state->fail_after_count >= 0) {                        \
+            if (state->fail_after_count-- == 0) {                  \
```
Comment (serhiy-storchaka) [inline Modules/_sre/sre_lib.h:575]: Setting `state->fail_after_count` does not affect subsequent calls of the `Pattern` methods, but it still affects the `finditer()` iterator which reuses state between iterations.

Comment (wjssz) [reply Modules/_sre/sre_lib.h:575]: I deleted my previous reply. 
Maybe setting `state->fail_after_count = -1` doesn't make sense, it's better to make the `finditer()` iterator stop the subsequent matchings.

**Comment by wjssz** [comment]
I remember another thing, if this patch is going to be backported, you may remove the changes related to this:
```diff
-    PyObject* pattern;	
+    PatternObject* pattern;
```
The patch will be smaller and clearer.
Polishing work can be done only on the `main` branch, such as change the type, remove the outdated `SRE_ERROR_RECURSION_LIMIT`.

**Comment by serhiy-storchaka** [comment]
> I remember another thing, if this patch is going to be backported, you may remove the changes related to this:

It is easier to not do. I can introduce new bugs when trying to remove these changes in backports.

Initially it was needed to pass the second argument to state_fini, after reverting that change it is just a clean up, but it can still be helpful if in future we will need to pass the second argument to state_fini.

**Comment by wjssz** [comment]
`sre` only has the “`\B` can't match empty string” bug as far as I know. It's easy to fix, if decide to fix it.
Maybe people won't touch `sre`'s code for many years.

**Pull request merged**